include/nbl/asset/ICPUDescriptorSet.h

```diff
@@ -144,57 +171,39 @@ class NBL_API ICPUDescriptorSet final : public IDescriptorSet<ICPUDescriptorSetL
 		}
 		inline const ICPUDescriptorSetLayout* getLayout() const { return m_layout.get(); }
 
-		//!
-		inline uint32_t getMaxDescriptorBindingIndex() const
+		std::pair<core::SRange<core::smart_refctd_ptr<IDescriptor>>, core::SRange<SDescriptorInfo::SBufferImageInfo>> getDescriptors(const uint32_t binding)
 		{
-			return m_bindingInfo ? static_cast<uint32_t>(m_bindingInfo->size()-1u):0u;
-		}
+			// TODO(achal): We are doing multiple binary searches here, albeit over different arrays, but still this can be avoided
+			// if we store more data in SBindingRedirect, specifically the type and the count.
+			const auto bindingInfo = std::lower_bound(getLayout()->getBindings().begin(), getLayout()->getBindings().end(), ICPUDescriptorSetLayout::SBinding{binding});
+			assert(bindingInfo->binding == binding && "binding is not in the descriptor set!");
```

Comment:
Even better, only use `IDescriptorSetLayout::SBinding` for speccing out the layout (basically only as a parameter for the constructor) and use `IDescriptorSetLayout::SBindingRedirect` (after appropriate renaming) to store all the stuff you need to remember

Reply:
> I thought long and hard about how to avoid this, but the best I came up with was, as the comment mentions, keeping more data in `SBindingRedirect`, specifically the `asset::E_DESCRIPTOR_TYPE`.
> 
> If we store the descriptor count as well, we can also avoid doing [this computation](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5ae62476fb0d73ac3ea4ce8e4a59ab21e8146483/include/nbl/video/utilities/IGPUObjectFromAssetConverter.h#L1870-L1876) (but this case might not be as bad).
> 
> If we decide to do both of the above, now we have lots of duplicate data (binding numbers, types and counts) in `IDescriptorSetLayout::SBindingRedirect` and `IDescriptorSetLayout::SBinding`, so my final idea is to merge both of them into one. In other words, segregate [`IDescriptorSetLayout::m_bindings`](https://github.com/Devsh-Graphics-Programming/Nabla/blob/5ae62476fb0d73ac3ea4ce8e4a59ab21e8146483/include/nbl/asset/IDescriptorSetLayout.h#L312) based on `asset::E_DESCRIPTOR_TYPE` and associate the redirect offsets with them to make a `SBindingRedirect` equivalent, and of course get rid of `SBindingRedirect` entirely, then.

yes briliant idea, sorry I took so long to fully grasp it.

Its basically the same thing as the things I focused on during review of `IDescriptorSetLayout`.

The only thing extra I'd suggest is to use `inclusive_sum` for the offsets as `offset[LAST]` will now contain the total number of descriptors, and before `offset[0]` was wasted anyway as it was always equal to 0